File: my-folder/problems/3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        set<vector<int>> ansset;
        int j,k, target;
        
        sort(nums.begin(), nums.end());
        
        for (int i=0; i<nums.size()-2; i++){
            
            j = i+1, k = nums.size() - 1;
            target = -1*(nums[i]);
            while (j < k){
//                 while (j+1 < k && nums[j] == nums[j+1]){
//                     j++;
//                 }
                
//                 while (k-1 > j && nums[k] == nums[k-1]){
//                     k--;
//                 }
                
                if (nums[j] + nums[k] == target){
                    ansset.insert({nums[i], nums[j], nums[k]});
                    j++, k--;
                }
                else if (nums[j] + nums[k] < target){
                    j++;
                }
                else{
                    k--;
                }
            }
        }
        
        for (auto v: ansset){
            ans.push_back(v);
        }
        
        return ans;
    }
};

File: my-folder/problems/3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        ans = []
        
        nums.sort()
        
        for i in range(0,len(nums)):
            if i>0 and nums[i-1] == nums[i]:
                continue
            
            j = i+1
            k = len(nums) - 1
            
            while j<k:
                '''while j+1<k and nums[j] == nums[j+1]:
                    j = j+1
                while j<k-1 and nums[k] == nums[k-1]:
                    k = k-1'''
                
                if nums[i] + nums[j] + nums[k] == 0:
                    ans.append([nums[i],nums[j],nums[k]])
                    while j<k and nums[j] == nums[j+1]:
                        j += 1
                    while j<k and nums[k] == nums[k-1]:
                        k -= 1
                    j += 1
                    k -= 1
                elif nums[i] + nums[j] + nums[k] > 0:
                    k -= 1
                else:
                    j += 1
        
        return ans